### PR TITLE
qt: depend on Python 3

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

**Comment by maxim-belkin** [comment]
My suggestions:
1. Replace `linuxbrew/xorg/mesa` with `linuxbrew/xorg/libglvnd`
2. Replace `linuxbrew/xorg/xorg` with:
  + depends_on "linuxbrew/xorg/libdrm"
  + depends_on "linuxbrew/xorg/libice"
  + depends_on "linuxbrew/xorg/libsm"
  + depends_on "linuxbrew/xorg/libxcomposite"
  + depends_on "linuxbrew/xorg/wayland"
  + whatever package that provides `libgbm.so.1`

**Comment by jonchang** [comment]
>     * whatever package that provides `libgbm.so.1`

this appears to be `mesa`? https://packages.debian.org/sid/libgbm-dev

**Comment by maxim-belkin** [comment]
`inreplace` is not going to work. `qt` has got some Python 2 code -- Python 3 can't process it.

> this appears to be mesa? https://packages.debian.org/sid/libgbm-dev

Hmm...

**Comment by jonchang** [comment]
> `inreplace` is not going to work. `qt` has got some Python 2 code -- Python 3 can't process it.

basing this off of the link from this comment: https://github.com/Homebrew/linuxbrew-core/issues/19644#issuecomment-588540403

I can't test this locally because my docker instance ran out of disk space, lol. Have to figure out how to configure that.

**Comment by maxim-belkin** [comment]
:) You got almost all the fixes I found. The last ones are:
1. you need python 3 ( hint: inreplace suggests that)
2. add bison, flex, gperf, and replace `xorg` with direct dependencies
  +    depends_on "linuxbrew/xorg/libdrm"
  +    depends_on "linuxbrew/xorg/libice"
  +    depends_on "linuxbrew/xorg/libsm"
  +    depends_on "linuxbrew/xorg/libxcomposite"
  +    depends_on "linuxbrew/xorg/wayland"
  +    depends_on "bison"
  +    depends_on "flex"
  +    depends_on "gperf"

File: Formula/qt.rb [hunk 1]
```diff
@@ -26,21 +26,30 @@ class Qt < Formula
...
+    depends_on "linuxbrew/xorg/wayland"
+    depends_on "linuxbrew/xorg/xcb-util"
     depends_on "linuxbrew/xorg/xcb-util-image"
     depends_on "linuxbrew/xorg/xcb-util-keysyms"
     depends_on "linuxbrew/xorg/xcb-util-renderutil"
-    depends_on "linuxbrew/xorg/xcb-util"
     depends_on "linuxbrew/xorg/xcb-util-wm"
-    depends_on "linuxbrew/xorg/xorg"
+    depends_on "pulseaudio"
+    depends_on "python@3"
```
Comment (iMichka) [inline Formula/qt.rb]: "python@3.8" ?

Comment (jonchang) [reply Formula/qt.rb]: time for me to take a break 🤦‍♂️

**Comment by jonchang** [comment]
Lots of build failures but I think this is the root cause

```
22423:2020-02-21T09:23:51.7134389Z /tmp/qt-20200221-31545-1s0zyzp/qt-everywhere-src-5.14.1/qtbase/include/QtCore/../../src/corelib/serialization/qxmlstream.h:528:2: fatal error: can’t write PCH file: No space left on device
```

**Comment by iMichka** [comment]
We should have 14 GB of SSD disk space. Looks like this is not enough ...

**Comment by jonchang** [comment]
I suspect the VM has space but docker isn’t allocated enough. We’ll need to find some way of giving docker more disk space.

**Comment by iMichka** [comment]
I tried to build this locally in our docker container:
```
jobs/abstractpickingjob.cpp: In static member function ‘static Qt3DRender::RayCasting::QRay3D Qt3DRender::Render::AbstractPickingJob::intersectionRay(const QPoint&, const Matrix4x4&, const Matrix4x4&, const QRect&)’:
jobs/abstractpickingjob.cpp:111:71: error: no matching function for call to ‘Qt3DCore::Vector3D_SSE::unproject(const Matrix4x4&, const Matrix4x4&, const QRect&)’
     nearPos = nearPos.unproject(viewMatrix, projectionMatrix, viewport);
                                                                       ^
In file included from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector3d_sse_p.h:1:0,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_p.h:61,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector3d_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector4d_sse_p.h:54,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector4d_sse_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector4d_p.h:61,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector4d_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/matrix4x4_sse_p.h:54,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/matrix4x4_sse_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/matrix4x4_p.h:70,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/matrix4x4_p.h:1,
                 from ../../include/Qt3DRender/5.14.1/Qt3DRender/private/../../../../../src/render/raycasting/qray3d_p.h:56,
                 from ../../include/Qt3DRender/5.14.1/Qt3DRender/private/qray3d_p.h:1,
                 from jobs/abstractpickingjob_p.h:55,
                 from jobs/abstractpickingjob.cpp:40:
../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_sse_p.h:182:42: note: candidate: Qt3DCore::Vector3D_SSE Qt3DCore::Vector3D_SSE::unproject(const Qt3DCore::Matrix4x4_AVX2&, const Qt3DCore::Matrix4x4_AVX2&, const QRect&) const
     Q_3DCORE_PRIVATE_EXPORT Vector3D_SSE unproject(const Matrix4x4_AVX2 &modelView, const Matrix4x4_AVX2 &projection, const QRect &viewport) const;
                                          ^
../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_sse_p.h:182:42: note:   no known conversion for argument 1 from ‘const Matrix4x4 {aka const Qt3DCore::Matrix4x4_SSE}’ to ‘const Qt3DCore::Matrix4x4_AVX2&’
jobs/abstractpickingjob.cpp:113:69: error: no matching function for call to ‘Qt3DCore::Vector3D_SSE::unproject(const Matrix4x4&, const Matrix4x4&, const QRect&)’
     farPos = farPos.unproject(viewMatrix, projectionMatrix, viewport);
                                                                     ^
In file included from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector3d_sse_p.h:1:0,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_p.h:61,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector3d_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector4d_sse_p.h:54,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector4d_sse_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector4d_p.h:61,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/vector4d_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/matrix4x4_sse_p.h:54,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/matrix4x4_sse_p.h:1,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/matrix4x4_p.h:70,
                 from ../../include/Qt3DCore/5.14.1/Qt3DCore/private/matrix4x4_p.h:1,
                 from ../../include/Qt3DRender/5.14.1/Qt3DRender/private/../../../../../src/render/raycasting/qray3d_p.h:56,
                 from ../../include/Qt3DRender/5.14.1/Qt3DRender/private/qray3d_p.h:1,
                 from jobs/abstractpickingjob_p.h:55,
                 from jobs/abstractpickingjob.cpp:40:
../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_sse_p.h:182:42: note: candidate: Qt3DCore::Vector3D_SSE Qt3DCore::Vector3D_SSE::unproject(const Qt3DCore::Matrix4x4_AVX2&, const Qt3DCore::Matrix4x4_AVX2&, const QRect&) const
     Q_3DCORE_PRIVATE_EXPORT Vector3D_SSE unproject(const Matrix4x4_AVX2 &modelView, const Matrix4x4_AVX2 &projection, const QRect &viewport) const;
                                          ^
../../include/Qt3DCore/5.14.1/Qt3DCore/private/../../../../../src/core/transforms/vector3d_sse_p.h:182:42: note:   no known conversion for argument 1 from ‘const Matrix4x4 {aka const Qt3DCore::Matrix4x4_SSE}’ to ‘const Qt3DCore::Matrix4x4_AVX2&’
```

**Comment by jonchang** [comment]
Can we try enabling AVX2? Looks like Qt doesn’t test all combinations of its options, lol

**Comment by iMichka** [comment]
Yes, I am rebuild it with AVX2 on now. I needed to disable it back then for another error ...

**Comment by iMichka** [comment]
```
g++ -c -pipe -O3 -std=c++1y -fvisibility=hidden -fvisibility-inlines-hidden -D_REENTRANT -fno-exceptions -Wall -Wextra -Wvla -Wdate-time -fPIC -DQT_NO_LINKED_LIST -DQT_NO_JAVA_STYLE_ITERATORS -DQT_NO_USING_NAMESPACE -DQT_NO_FOREACH -DENABLE_PIXMAN_DRAWHELPERS -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_BUILD_GUI_LIB -DQT_BUILDING_QT -DQT_NO_CAST_TO_ASCII -DQT_ASCII_CAST_WARNINGS -DQT_MOC_COMPAT -DQT_USE_QSTRINGBUILDER -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_BEFORE=0x050000 -DQT_DEPRECATED_WARNINGS_SINCE=0x060000 -DQT_NO_DEBUG -DQT_CORE_LIB -DQT_CORE_LIB -DQT_NO_EXCEPTIONS -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I. -I../3rdparty/md4c -I../3rdparty/zlib/src -I../3rdparty/VulkanMemoryAllocator -I../../include -I../../include/QtGui -I../../include/QtGui/5.14.1 -I../../include/QtGui/5.14.1/QtGui -I.tracegen -I/home/linuxbrew/.linuxbrew/Cellar/libxau/1.0.9/include -I/home/linuxbrew/.linuxbrew/Cellar/libxdmcp/1.1.3/include -I/home/linuxbrew/.linuxbrew/Cellar/libxcb/1.13.1/include -I/home/linuxbrew/.linuxbrew/Cellar/libdrm/2.4.100/include -I/home/linuxbrew/.linuxbrew/Cellar/libdrm/2.4.100/include/libdrm -I/home/linuxbrew/.linuxbrew/Cellar/libx11/1.6.9/include -I/home/linuxbrew/.linuxbrew/Cellar/libxfixes/5.0.3/include -I/home/linuxbrew/.linuxbrew/Cellar/libxext/1.3.4/include -I/home/linuxbrew/.linuxbrew/Cellar/libxdamage/1.1.5/include -I/home/linuxbrew/.linuxbrew/Cellar/libxxf86vm/1.1.4/include -I/home/linuxbrew/.linuxbrew/Cellar/mesa/19.3.3/include -I/home/linuxbrew/.linuxbrew/Cellar/xorgproto/2019.2/include -I../../include/QtCore/5.14.1 -I../../include/QtCore/5.14.1/QtCore -I../../include/QtCore -I.moc -I../3rdparty/libpng -I../3rdparty/harfbuzz-ng/include -I../../mkspecs/linux-g++ image/qimage_compat.cpp -o .obj/qimage_compat.o
2020-02-25T10:23:08.8144776Z painting/qdrawhelper_avx2.cpp: In function â€˜void BYTE_MUL_AVX2(__m256i&, __m256i, __m256i, __m256i)â€™:
2020-02-25T10:23:08.8149706Z painting/qdrawhelper_avx2.cpp:59:1: note: The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6
2020-02-25T10:23:08.8149968Z  BYTE_MUL_AVX2(__m256i &pixelVector, __m256i alphaChannel, __m256i colorMask, __m256i half)
2020-02-25T10:23:08.8151401Z  ^
2020-02-25T10:23:08.8153347Z In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include/immintrin.h:43:0,
2020-02-25T10:23:08.8156353Z                  from /usr/lib/gcc/x86_64-linux-gnu/5/include/x86intrin.h:46,
2020-02-25T10:23:08.8159304Z                  from /usr/include/x86_64-linux-gnu/c++/5/bits/opt_random.h:33,
2020-02-25T10:23:08.8159467Z                  from /usr/include/c++/5/random:50,
2020-02-25T10:23:08.8159575Z                  from /usr/include/c++/5/bits/stl_algo.h:66,
2020-02-25T10:23:08.8159678Z                  from /usr/include/c++/5/algorithm:62,
2020-02-25T10:23:08.8161103Z                  from ../../include/QtCore/../../src/corelib/global/qglobal.h:142,
2020-02-25T10:23:08.8164210Z                  from ../../include/QtCore/qglobal.h:1,
2020-02-25T10:23:08.8164358Z                  from ../../include/QtGui/../../src/gui/kernel/qtguiglobal.h:43,
2020-02-25T10:23:08.8166169Z                  from ../../include/QtGui/qtguiglobal.h:1,
2020-02-25T10:23:08.8167602Z                  from ../../include/QtGui/5.14.1/QtGui/private/../../../../../src/gui/kernel/qtguiglobal_p.h:54,
2020-02-25T10:23:08.8170644Z                  from ../../include/QtGui/5.14.1/QtGui/private/qtguiglobal_p.h:1,
2020-02-25T10:23:08.8170749Z                  from painting/qdrawhelper_p.h:54,
2020-02-25T10:23:08.8172557Z                  from painting/qdrawhelper_avx2.cpp:41:
2020-02-25T10:23:08.8202942Z /usr/lib/gcc/x86_64-linux-gnu/5/include/avx2intrin.h:574:1: error: inlining failed in call to always_inline â€˜__m256i _mm256_or_si256(__m256i, __m256i)â€™: target specific option mismatch
2020-02-25T10:23:08.8203109Z  _mm256_or_si256 (__m256i __A, __m256i __B)
2020-02-25T10:23:08.8203416Z  ^
2020-02-25T10:23:08.8203521Z painting/qdrawhelper_avx2.cpp:75:34: error: called from here
2020-02-25T10:23:08.8203616Z      pixelVector = _mm256_or_si256(pixelVectorAG, pixelVectorRB);
2020-02-25T10:23:08.8204098Z                                   ^
2020-02-25T10:23:08.8204581Z In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include/immintrin.h:43:0,
2020-02-25T10:23:08.8204915Z                  from /usr/lib/gcc/x86_64-linux-gnu/5/include/x86intrin.h:46,
2020-02-25T10:23:08.8205406Z                  from /usr/include/x86_64-linux-gnu/c++/5/bits/opt_random.h:33,
2020-02-25T10:23:08.8205529Z                  from /usr/include/c++/5/random:50,
2020-02-25T10:23:08.8205634Z                  from /usr/include/c++/5/bits/stl_algo.h:66,
2020-02-25T10:23:08.8205736Z                  from /usr/include/c++/5/algorithm:62,
2020-02-25T10:23:08.8205829Z                  from ../../include/QtCore/../../src/corelib/global/qglobal.h:142,
2020-02-25T10:23:08.8205939Z                  from ../../include/QtCore/qglobal.h:1,
2020-02-25T10:23:08.8206191Z                  from ../../include/QtGui/../../src/gui/kernel/qtguiglobal.h:43,
2020-02-25T10:23:08.8206302Z                  from ../../include/QtGui/qtguiglobal.h:1,
2020-02-25T10:23:08.8206544Z                  from ../../include/QtGui/5.14.1/QtGui/private/../../../../../src/gui/kernel/qtguiglobal_p.h:54,
2020-02-25T10:23:08.8206665Z                  from ../../include/QtGui/5.14.1/QtGui/private/qtguiglobal_p.h:1,
2020-02-25T10:23:08.8206775Z                  from painting/qdrawhelper_p.h:54,
2020-02-25T10:23:08.8235298Z                  from painting/qdrawhelper_avx2.cpp:41:
2020-02-25T10:23:08.8236426Z /usr/lib/gcc/x86_64-linux-gnu/5/include/avx2intrin.h:186:1: error: inlining failed in call to always_inline â€˜__m256i _mm256_andnot_si256(__m256i, __m256i)â€™: target specific option mismatch
2020-02-25T10:23:08.8236857Z  _mm256_andnot_si256 (__m256i __A, __m256i __B)
```

**Comment by iMichka** [comment]
See https://bugreports.qt.io/browse/QTBUG-71564. This is why I disabled avx2 back then.

**Comment by iMichka** [comment]
Let's try with gcc7, just in case this helps.

**Comment by iMichka** [comment]
```
Filesystem      Size  Used Avail Use% Mounted on
overlay          84G   66G   18G  79% /
tmpfs            64M     0   64M   0% /dev
tmpfs           3.4G     0  3.4G   0% /sys/fs/cgroup
shm              64M     0   64M   0% /dev/shm
/dev/sda1        84G   66G   18G  79% /__t
tmpfs           695M  768K  694M   1% /run/docker.sock
tmpfs           3.4G     0  3.4G   0% /proc/acpi
tmpfs           3.4G     0  3.4G   0% /proc/scsi
tmpfs           3.4G     0  3.4G   0% /sys/firmware
```

**Comment by maxim-belkin** [comment]
```
2020-02-29T21:41:54.4697787Z Info: creating super cache file /tmp/qt-20200229-28792-1wztlkj/qt-everywhere-src-5.14.1/.qmake.super
2020-02-29T21:41:54.4698376Z Info: creating cache file /tmp/qt-20200229-28792-1wztlkj/qt-everywhere-src-5.14.1/.qmake.cache
2020-02-29T21:41:54.4698909Z ERROR: Invalid value 'demos' supplied to command line option 'nomake'.
2020-02-29T21:41:54.4699050Z 
2020-02-29T21:41:54.4699478Z ERROR: Unknown command line option '-no-gtkstyle'.
2020-02-29T21:42:05.1351287Z 
2020-02-29T21:42:05.1353773Z [34m==>[0m [1mFormula[0m
2020-02-29T21:42:05.1354385Z Path: /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/qt.rb
2020-02-29T21:42:05.1354899Z [34m==>[0m [1mConfiguration[0m
```

**Comment by iMichka** [comment]
I'm trying to disable some options to reduce disk space usage. But this is only a short term solution.

Maybe I could add a `brew cleanup --prune=0` at the beginning of the install step? This is what test-bot does when disk space gets small. We might need to do this earlier for this formula.

**Comment by iMichka** [comment]
Merged as-is, I made a follow up commit: 6418315fa35ae082f7a470df1e21837932802386

**Pull request merged**